stream_alert/rule_processor/rules_engine.py

```diff
@@ -374,24 +376,46 @@ def process(cls, input_payload):
                 if types_result:
                     record_copy = record.copy()
                     record_copy[NORMALIZATION_KEY] = types_result
+                    if StreamThreatIntel.enabled():
+                        cls.__normalized_records.append(record_copy)
                 else:
                     record_copy = record
                 # rule analysis
-                rule_result = cls.process_rule(record_copy, rule)
-                if rule_result:
-                    LOGGER.info('Rule [%s] triggered an alert on log type [%s] from entity \'%s\' '
-                                'in service \'%s\'', rule.rule_name, payload.log_source,
-                                payload.entity, payload.service())
-                    alert = {
-                        'record': record_copy,
-                        'rule_name': rule.rule_name,
-                        'rule_description': rule.rule_function.__doc__ or DEFAULT_RULE_DESCRIPTION,
-                        'log_source': str(payload.log_source),
-                        'log_type': payload.type,
-                        'outputs': rule.outputs,
-                        'source_service': payload.service(),
-                        'source_entity': payload.entity,
-                        'context': rule.context}
-                    alerts.append(alert)
+                cls.rule_analysis(record_copy, rule, payload, alerts)
+
+        # Apply the Threat Intelligence against normalized records
+        if StreamThreatIntel.enabled():
+            threat_intel = StreamThreatIntel()
+            ioc_records = threat_intel.threat_detection(cls.__normalized_records)
+            if ioc_records:
+                for ioc_record in ioc_records:
+                    for rule in rules:
+                        cls.rule_analysis(ioc_record, rule, payload, alerts)
 
         return alerts
+
+    @classmethod
+    def rule_analysis(cls, record, rule, payload, alerts):
+        """Static method to analyze rule
```

Comment:
nit: I think it's ok to leave out `static method` from here and just describe the main use case of the function 😄